### PR TITLE
Documentation: Add a document about the LibJS garbage collection

### PR DESCRIPTION
**Pull request by flofriday**
This document describes how the GC works at the moment, how it integrates with the Interpreter and the intentions behind some of the classes that make up the GC.

### CODE REVIEW

File: Documentation/Browser/LibJSGarbageCollection.md
Comment (elisee) [inline Documentation/Browser/LibJSGarbageCollection.md:7]:
```suggestion
Allocation and garbage collection are tightly integrated and therefore the story of a JS object and its relationship with the GC starts at its allocation. The `Interpreter` has a member of the `Heap` class which allows to allocate a new `Cell` which will be kept alive as long as the Interpreter needs it (has an active reference to it). While the GC is an integral part of the JS engine, it is not limited to just JS objects. Any class like the `Object` class (JS object) can inherit from `Cell` and can then be managed by the GC. The browser for example also leverages the GC to manage DOM elements.
```

File: Documentation/Browser/LibJSGarbageCollection.md
Comment (elisee) [inline Documentation/Browser/LibJSGarbageCollection.md:9]:
```suggestion
The actual allocations of the objects aren't made with `malloc` and `free` but are made inside `HeapBlock`s of 16KB each. This is a performance improvement as the GC has to visit the entire heap during collection and this increases locality (without being a compacting GC).
```

File: Documentation/Browser/LibJSGarbageCollection.md
Comment (elisee) [inline Documentation/Browser/LibJSGarbageCollection.md:11]:
```suggestion
The `Heap` class outsources the allocations to the `CellAllocator` which manages a list of heap blocks. Cells can have different sizes because different JS objects inherit from `Cell`. Therefore,  multiple cell allocators exist where each one manages cells of a particular maximum size and the smallest allocator that can hold the required cell is chosen.
```

File: Documentation/Browser/LibJSGarbageCollection.md
Comment (elisee) [inline Documentation/Browser/LibJSGarbageCollection.md:13]:
```suggestion
In case all available heap blocks of a cell allocator are full, a new heap block is allocated by the `BlockAllocator`. The block allocator is a general class to allocate blocks of memory and is not bound to the `HeapBlock` class. The big advantage of this class is that it doesn't immediately free empty blocks but keeps them around for some time, in case it can reuse them ([relevant Andreas Kling video: Improving JS allocation performance by recycling GC heap blocks](https://www.youtube.com/watch?v=x9xLOuMK9wo)). If no available heap blocks are available, a new one will be allocated with `aligned_alloc`.
```

File: Documentation/Browser/LibJSGarbageCollection.md
Comment (elisee) [inline Documentation/Browser/LibJSGarbageCollection.md:19]:
```suggestion
There is also an option to stress-test the GC and run it on every allocation. (`js --gc-on-every-allocation` or `js -g` when running the standalone interpreter)
```

File: Documentation/Browser/LibJSGarbageCollection.md
Comment (elisee) [inline Documentation/Browser/LibJSGarbageCollection.md:23]:
```suggestion
To mark all reachable cells, an initial set of cells is needed which is often called the GC-roots. LibJS gathers these roots from the `VM` and also with a conservative GC that scans all registers and the stack for pointers that _could_ reference a `Cell`. This was added so that the GC doesn't remove an object while it is being constructed which might take multiple allocations ([relevant Andreas Kling video: Conservative garbage collection](https://www.youtube.com/watch?v=c6IghR0W254)).
```

File: Documentation/Browser/LibJSGarbageCollection.md
Comment (elisee) [inline Documentation/Browser/LibJSGarbageCollection.md:25]:
```suggestion
Cells can hold references to other Cells and during marking the GC needs to trace them all. This is solved with the `visit_edges` virtual method that classes inheriting from `Cell` must implement.
```

**Comment by flofriday** [comment]
Hey, is there anything else I should change? I am just a little lost about how this gets merged. Will you eventually set the `pr-community-approved` label or are we waiting for another reviewer for a second opinion?

Of course i don't wanna pressure you, if you just need more time. ;)

**Comment by gmta** [comment]
Hi @flofriday, eventually a maintainer (such as myself, but ideally someone with more LibJS knowledge) will give it a review. The `pr-community-approved` label doesn't really do a lot in terms of mergeability, it's just a hint that someone already looked at it. Optionally you can promote your PR in #code-review on Discord!

**Review by elisee — APPROVED** [review]

**Comment by ronak69** [comment]
Add the new doc in `Documentation/README.md`.

**Pull request closed without merging**
